File: 05월 24일 - 유니온 파인드/과제/12100.cpp
#include <iostream>
#include <vector>

using namespace std;

int n, ans = 0;
vector<vector<int>> board;

void print() {
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            cout << board[i][j] << ' ';
        }
        cout << '\n';
    }
}

void upMove() {
    vector<vector<int>> temp(n, vector<int>(n, 0));
    for (int j = 0; j < n; j++) {
        int cnt = 0;
        for (int i = 0; i < n; i++) {
            if (board[i][j]) {
                temp[cnt++][j] = board[i][j];
            }
        }
    }
    board = temp;
    temp.assign(n, vector<int>(n, 0));
    for (int j = 0; j < n; j++) {
        int cnt = 0;
        for (int i = 0; i < n - 1; i++) {
            if (board[i][j] && board[i][j] == board[i + 1][j]) {
                board[i + 1][j] = 0;
                temp[cnt++][j] = board[i][j] * 2;
                i++;
            } else if (board[i][j]) {
                temp[cnt++][j] = board[i][j];
            }
        }
        if (board[n - 1][j]) {
            temp[cnt++][j] = board[n - 1][j];
        }
    }
    board = temp;
}

void downMove() {
    vector<vector<int>> temp(n, vector<int>(n, 0));
    for (int j = 0; j < n; j++) {
        int cnt = n - 1;
        for (int i = n - 1; i >= 0; i--) {
            if (board[i][j]) {
                temp[cnt--][j] = board[i][j];
            }
        }
    }
    board = temp;
    temp.assign(n, vector<int>(n, 0));
    for (int j = 0; j < n; j++) {
        int cnt = n - 1;
        for (int i = n - 1; i > 0; i--) {
            if (board[i][j] && board[i][j] == board[i - 1][j]) {
                board[i - 1][j] = 0;
                temp[cnt--][j] = board[i][j] * 2;
                i--;
            } else if (board[i][j]) {
                temp[cnt--][j] = board[i][j];
            }
        }
        if (board[0][j]) {
            temp[cnt--][j] = board[0][j];
        }
    }
    board = temp;
}

void leftMove() {
    vector<vector<int>> temp(n, vector<int>(n, 0));
    for (int i = 0; i < n; i++) {
        int cnt = 0;
        for (int j = 0; j < n; j++) {
            if (board[i][j]) {
                temp[i][cnt++] = board[i][j];
            }
        }
    }
    board = temp;
    temp.assign(n, vector<int>(n, 0));
    for (int i = 0; i < n; i++) {
        int cnt = 0;
        for (int j = 0; j < n - 1; j++) {
            if (board[i][j] && board[i][j] == board[i][j + 1]) {
                board[i][j + 1] = 0;
                temp[i][cnt++] = board[i][j] * 2;
                j++;
            } else if (board[i][j]) {
                temp[i][cnt++] = board[i][j];
            }
        }
        if (board[i][n - 1]) {
            temp[i][cnt++] = board[i][n - 1];
        }
    }
    board = temp;
}

void rightMove() {
    vector<vector<int>> temp(n, vector<int>(n, 0));
    for (int i = 0; i < n; i++) {
        int cnt = n - 1;
        for (int j = n - 1; j >= 0; j--) {
            if (board[i][j]) {
                temp[i][cnt--] = board[i][j];
            }
        }
    }
    board = temp;
    temp.assign(n, vector<int>(n, 0));
    for (int i = 0; i < n; i++) {
        int cnt = n - 1;
        for (int j = n - 1; j > 0; j--) {
            if (board[i][j] && board[i][j] == board[i][j - 1]) {
                board[i][j - 1] = 0;
                temp[i][cnt--] = board[i][j] * 2;
                j--;
            } else if (board[i][j]) {
                temp[i][cnt--] = board[i][j];
            }
        }
        if (board[i][0]) {
            temp[i][cnt--] = board[i][0];
        }
    }
    board = temp;
}

int maxBlock() {
    int max_block = 0;
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            max_block = max(max_block, board[i][j]);
        }
    }
    return max_block;
}

void backtracking(int cnt) {
    //print();
    if (cnt == 5) {
        ans = max(ans, maxBlock());
        return;
    }
    //상, 하, 좌, 우로 모두 이동시켜보기
    vector<vector<int>> temp = board;
    upMove();
    backtracking(cnt + 1);
    board = temp;

    downMove();
    backtracking(cnt + 1);
    board = temp;

    leftMove();
    backtracking(cnt + 1);
    board = temp;

    rightMove();
    backtracking(cnt + 1);
    board = temp;
}

/**
 * [2048]
 *
 * - 최대 5번을 이동시키는 모든 경우를 구한 후, 가장 큰 블록 찾는 문제 - 백트래킹 탐색
 * -
 */

int main() {
    //입력
    cin >> n;
    board.assign(n, vector<int>(n, 0));
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            cin >> board[i][j];
        }
    }

    //연산
    backtracking(0);

    //출력
    cout << ans;
    return 0;
}

File: 05월 24일 - 유니온 파인드/과제/12100.py
import sys
sys.setrecursionlimit(10**6)
input = sys.stdin.readline

"""
 [2048]

 - 최대 5번을 이동시키는 모든 경우를 구한 후, 가장 큰 블록 찾는 문제 - 백트래킹 탐색
 - 움직이는 함수는 하나만 짜고, 보드를 돌려가면서 상, 하, 좌, 우 모든 방향의 움직임을 만들 수 있다.

 - 상 <-> 하: 행 순서를 뒤집어서 해결
 - 상/하 <-> 좌/우: Transpose Matrix를 만들면 된다.

 - zip(): 두개 이상의 iterable한 객체를 인자로 받아서, 순서대로 하나씩 묶어서 반환한다.
    ex) zip([1, 2, 3], ['a', 'b', 'c']) -> [1, 'a'], [2, 'b'], [3, 'c']
    따라서 zip의 인자로 각 행을 넣어주면, 행과 열이 바뀐 Transpose matrix가 나오게 된다.
"""

def up(board):
    temp = [[0]*n for _ in range(n)]
    for c in range(n):
        idx = 0
        prev = 0
        for r in range(n):
            if not board[r][c]:
                continue
            if board[r][c] == prev:
                temp[idx-1][c] *= 2
                prev = 0
            else:
                temp[idx][c] = board[r][c]
                prev = board[r][c]
                idx += 1

    return temp

def backtracking(cnt, board):
    global ans
    if cnt == 5:
        for line in board:
            ans = max(ans, max(line))
        return

    # Transpose matrix 구하기 (상 -> 좌 / 하 -> 우)
    board_t = list(zip(*board))

    # 위로
    backtracking(cnt+1, up(board))
    # 아래로
    backtracking(cnt+1, up(board[::-1]))
    # 왼쪽으로
    backtracking(cnt+1, up(board_t))
    # 오른쪽으로
    backtracking(cnt+1, up(board_t[::-1]))

    return

# 입력
n = int(input())
board = [list(map(int, input().split())) for _ in range(n)]

ans = 0
backtracking(0, board)

# 출력
print(ans)